Create an unbundled route for NPM users of Lit

KonnorRogers opened the pull request:
instead of 

Fixes #559 

Currently, imports look like this: 

```js
import "@shoelace-style/shoelace/dist/components/button/button.js"
```

With this PR there'd be an additional route for npm under "dist"

```js
import "@shoelace-style/shoelace/dist/npm/components/button/button.js"
```

The other option is to make another top-level route and expose it in `files: ["npm"]`

and then you'd import like this:

```js
import "@shoelace-style/shoelace/npm/components/button/button.js"
```

Alternatively, we could solve the extra exports with exportmaps, kind of open to any ideas, each come with caveats, but my goal of the PR was to not break existing builds and CDN Routes.

KonnorRogers committed:
build script for NPM deployments

KonnorRogers committed:
import path

KonnorRogers committed:
prettier

claviska committed:
wait until registered to set initial state; fixes #1292

claviska committed:
update changelog

claviska committed:
don't show hover when focused; fixes #1282

alenaksu committed:
fix(carousel): clickable elements don't work on chrome (#1266)

* fix(carousel): clickable elements don't work on chrome

* fix: update implementation

claviska committed:
update changelog

claviska committed:
less pipeline flakes 🤞🏻

dhellgartner committed:
Slot aria attributes (#1296)

* Fix acessability issue

* Additionally adapted the test

* Added more accessability tests

* Updated the testing documentation

to take the fact that accessability checks cover only
rendered content into account

---------

Co-authored-by: Dominikus Hellgartner <dominikus.hellgartner@gmail.com>

claviska committed:
fix typos

claviska committed:
revert role and don't use <header> for buttons

claviska committed:
no roles on slots; fixes #1287

mpharoah committed:
Improve performance of sl-icon by caching later (#1286)

* Improve performance of sl-icon by caching later

* Fixed error handling

* Don't use requestInclude in sl-icon

* Separate sl-icon errors into cacheable and retryable errors

claviska committed:
spell check and reorder static function

claviska commented:
> With this PR there'd be an additional route for npm under "dist"
> 
>     import "@shoelace-style/shoelace/dist/npm/components/button/button.js"

Thanks for jumpstarting this, @KonnorRogers!

Putting both builds into one package will probably confuse users, and I'd rather not support it long term if possible. Too many people browse `node_modules` and will inevitably import the wrong files, meaning we'll see duplicate imports, broken singletons, and other hard-to-diagnose problems.

IIRC, previous discussions with @justinfagnani indicated it would be better to ship the unbundled "npm version" of Shoelace through the existing package `@shoelace-style/shoelace` and then publish an _additional_ bundled package that we can serve from the CDN.*

I'm not sure if that would be `@shoelace-style/cdn` or what, but separate packages seems like the best idea (and I'm OK with breaking the CDN link in 2.0 to get this done).

At the same time, I'd like to avoid using Lerna or similar. But we can probably write a script to help automate the main release and, if successful, have it publish the CDN one too.

I know exactly what the CDN build requires, but I'm not sure what the unbundled npm version will require aside from, well, not bundling. @justinfagnani [shared some thoughts](https://github.com/shoelace-style/shoelace/issues/559#issuecomment-1209767132) and I want to make sure we capture them to make Shoelace friendlier for npm users.

I guess, at a minimum, we can create both packages (e.g. `@shoelace-style/shoelace` (unbundled) and `@shoelace-style/cdn` (bundled) and fill in the gaps, if any, once those are both live.

---

_*We previously discussed _not_ publishing a separate npm package for the CDN, but I think it will be necessary since 1) we don't commit the bundled `dist` to git and 2) many public CDN utilize npm as a source so things will "just work" if we do it that way. We can post in big letters in the README that users shouldn't import the CDN package via npm._

claviska commented:
Since it's fresh in my mind, a big reason I haven't done this yet was because one of the few metrics I have to hint at Shoelace's usage is jsDelivr's "hits per month," which this would temporarily break. (Shoelace is [currently the 65th most popular package](https://www.jsdelivr.com/package/npm/@shoelace-style/shoelace) by that metric.)

That's less of an issue to me now. We'll just add them together internally until things normalize.

claviska commented:
I forgot to mention that, at some point this year, we're going to be moving off the public CDN to a more reliable one powered by Font Awesome, so we probably don't want to lean on any jsDelivr-specific feature.

@supercodepoet do you know if our infra supports package exports, or should we keep the bundled CDN build simple and separate to avoid problems when we transition?

justinfagnani commented:
jsdelivr docs are hard to scan for infor about package exports. Apparently esm.run supports them, but I don't know what conditions they prefer. I also don't quite understand the relationship between jsdelivr and esm.run, since on an esm.run page if you open a package in a new tab you get a jsdelive URL, like `https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace/+esm` (that un-versioned URL scares me)

I noticed that all Shoelace modules have a `/dist` prefix. One option is to remove that for npm imports and not include `/dist` in any package exports. So npm users would do:

```ts
import '@shoelace-style/shoelace/shoelace.js';
```

(or just `import '@shoelace-style/shoelace'`)

and jsdeliver users would still do:

```html
<script type="module" src="https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.3.0/dist/shoelace.js"></script>
```

claviska commented:
> I noticed that all Shoelace modules have a /dist prefix. One option is to remove that for npm imports and not include /dist in any package exports. So npm users would do:

Wait, we don't want npm users to import the source files directly, do we? Wouldn't that require them to run _our_ version of TypeScript and all that?

Currently, npm users import a compiled JS version with dependencies bundled. I thought we wanted to keep doing that, but unbundle Lit and other ESM compatible libs in the process.

This PR appears to dump the unbundled build into `dist/npm`. I think we should probably publish it to `dist`, then create another folder (maybe `dist-cdn`?) and publish the _bundled CDN version_ there. Then we can release that as a separate repo for public CDNs, e.g. `@shoelace-style/cdn`.

So, TL;DR:

- output the unbundled npm version to `dist`
- output the bundled CDN version to `dist-cdn`
- DO NOT add `dist-cdn` to `package.json > files`
- publish `dist-cdn` as a separate npm package called `@shoelace-style/cdn`

Am I missing anything?

justinfagnani commented:
> Wait, we don't want npm users to import the source files directly, do we? Wouldn't that require them to run our version of TypeScript and all that?

They wouldn't import the source files, they'd import `.js`, but without the `/dist`/ prefix. You can either output the files to the root directory (that's what we generally do in Lit), or use the export map to point the importable specifier to files within `/dist` like `/dist/npm/shoelace.js`.

KonnorRogers commented:
Fwiw, if we're cool with breaking CDN, I would have a "/cdn"

So imports would look like this for CDN:

`import "@shoelace-style/shoelace/cdn/components/button/button.js"`

And then NPM imports would be left as is. The benefit of that too is it doesn't break typescript type checking.

marking dependencies as external should be all that shoelace needs to do

claviska commented:
I’m down with that. Let’s just not publish the CDN folder to npm. Or at least to a different package.

@KonnorRogers do you wanna take charge of this?

KonnorRogers commented:
@claviska i don't think there's a way to not publish a folder to NPM but still have it available on CDN.

claviska commented:
Separate repo would be OK. I’ll handle it manually for now, if need be. Just make it produce both folders and I’ll sort out publishing if you want.

jaredcwhite commented:
Just chiming in to say: I'm excited to see how this turns out!

KonnorRogers commented:
@claviska sounds good. I'll move things around in the PR to output to "./cdn" and "./dist" will be the canonical spot for NPM users. I'll make sure ./cdn is in .gitignore and is not part of the "files" key. Should be a relatively small change to what I have here.

claviska committed:
add size to radio group; fixes #1301

claviska committed:
fixes #1302

claviska committed:
bump version

claviska committed:
2.4.0

KonnorRogers committed:
generate a CDN path'

KonnorRogers committed:
shim for tests

KonnorRogers committed:
prettier

mpharoah committed:
Improve performance of sl-rating by not render all icons twice

mpharoah committed:
Fixed clipPath

claviska committed:
update docs; fixes #1315

claviska committed:
update radio size when group size changes

KonnorRogers committed:
feat: support variable height elements inside <sl-button>

claviska committed:
add checkbox + exported parts; #1318

claviska committed:
reorder properties

claviska committed:
update changelog

claviska committed:
update changelog; #1310

claviska committed:
add submenu-icon part

KonnorRogers committed:
tests: add regression tests for checkbox and toggle focus behavior (#1330)

* add regression test for checkbox focusing

* change number of checkboxes / switches

* change max-height to 400px so it fails

* re-add positon: relative;

Erick2280 committed:
Fix Portuguese translation strings (#1336)

yuki24 committed:
Add better support for Turbo Drive (#1338)

sba committed:
Update de.ts (#1339)

"Gleiten" is a verb and means glide, "Slide" must be translated with "Folie" (like in previousSlide, nextSlide)

claviska committed:
update changelog

claviska committed:
Remove outdated line from docs

Blackscr33n committed:
update angular getting started (#1352)

ErikOnBike committed:
Replace .map with .forEach when return value not used. See issue #740 (#1349)

claviska committed:
update form control guidelines

claviska committed:
reflect size; fixes #1348

dhellgartner committed:
Split panel tests (#1343)

* Added tests for sl-split-panel

test for horizontal arrangement

* Added tests for sl-split-panel

tests for vertical arrangement

---------

Co-authored-by: stefanie.hellgartner <stefanie.hellgartner@in-tech.com>

claviska committed:
update changelog

yuki24 committed:
Update the tutorial for Rails (#1258)

* Update the tutorial for Rails

* Update docs/tutorials/integrating-with-rails.md

Co-authored-by: Konnor Rogers <konnor5456@gmail.com>

* Update docs/tutorials/integrating-with-rails.md

Co-authored-by: Konnor Rogers <konnor5456@gmail.com>

* Update docs/tutorials/integrating-with-rails.md

Co-authored-by: Konnor Rogers <konnor5456@gmail.com>

* Update docs/tutorials/integrating-with-rails.md

Co-authored-by: Konnor Rogers <konnor5456@gmail.com>

* Update docs/tutorials/integrating-with-rails.md

Co-authored-by: Konnor Rogers <konnor5456@gmail.com>

* Update docs/tutorials/integrating-with-rails.md

* Add a bit more explanation to set up icons

---------

Co-authored-by: Konnor Rogers <konnor5456@gmail.com>

claviska committed:
update changelog

KonnorRogers commented:
@justinfagnani Could perhaps exportmaps be of use here?

```json
"exports": {
  ".": {
    "import" : "./dist/index.js",
    "browser": "./cdn/index.js"
  }
}
```

I don't know what the support status looks like of exportmaps for CDNs, but it seems like it'd keep us from having to publish a separate package.

justinfagnani commented:
I don't think export conditions help unless there's a CDN condition. You don't want to use "browser" because many bundlers will choose that condition and you'll get the pre-bundled Shoelace, with the bloat, lack of dev mode, lack of SSR support, etc. You need non-CDN environments to be able to choose browser, development, default, node, and combinations of as needed.

KonnorRogers commented on web-test-runner.config.js:
```diff
@@ -26,7 +26,9 @@ export default {
     <html lang="en-US">
-      <head></head>
+      <head>
+        <script>window.process = {env: { NODE_ENV: "production" }};</script>
```
Can't remember why this was needed 🤔

KonnorRogers committed:
remove duplication

KonnorRogers committed:
add a semicolon

KonnorRogers committed:
prettier

KonnorRogers commented:
Moving to: 

https://github.com/shoelace-style/shoelace/pull/1369
